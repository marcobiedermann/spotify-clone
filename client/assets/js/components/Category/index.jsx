import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import './style.css';

const Category = ({ id, icons, name }) => (
  <figure className="category">
    {icons[0] && (
      <Link to={`/browse/categories/${id}/playlists`}>
        <img src={icons[0].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <Link to={`/browse/categories/${id}/playlists`}>
        {name}
      </Link>
    </figcaption>
  </figure>
);

Category.propTypes = {
  id: PropTypes.string,
  icons: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
};

Category.defaultProps = {
  id: '',
  icons: [],
  name: '',
};

export default Category;
