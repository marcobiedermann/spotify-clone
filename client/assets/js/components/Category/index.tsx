import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import Image, { ImageProps } from '../Image';

export interface CategoryProps {
  icons: ImageProps[];
  id: string;
  name: string;
}

const Category: FC<CategoryProps> = (props) => {
  const { id, icons, name } = props;

  return (
    <figure>
      {icons[0] && (
        <Link to={`/browse/categories/${id}/playlists`}>
          <Image {...icons[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/browse/categories/${id}/playlists`}>{name}</Link>
      </figcaption>
    </figure>
  );
};

export default Category;
