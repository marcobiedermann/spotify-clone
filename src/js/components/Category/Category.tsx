import { Link } from 'react-router-dom';
import Image from '../Image';

const defaultIconSize = 100;

interface Icon {
  height: number | null;
  url: string;
  width: number | null;
}

export interface CategoryProps {
  icons: Icon[];
  id: string;
  name: string;
}

function Category(props: CategoryProps): JSX.Element {
  const { id, icons, name } = props;
  const icon = icons[0];

  return (
    <figure>
      {icons[0] && (
        <Link to={`/browse/categories/${id}/playlists`}>
          <Image
            {...icon}
            alt={name}
            width={icon.width || defaultIconSize}
            height={icon.height || defaultIconSize}
          />
        </Link>
      )}
      <figcaption>
        <Link to={`/browse/categories/${id}/playlists`}>{name}</Link>
      </figcaption>
    </figure>
  );
}

export default Category;
