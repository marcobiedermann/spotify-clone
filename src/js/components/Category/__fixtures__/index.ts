const icon = {
  height: 10,
  url: '',
  width: 100,
};

const category = {
  icons: [icon],
  id: '',
  name: '',
};

export default category;
