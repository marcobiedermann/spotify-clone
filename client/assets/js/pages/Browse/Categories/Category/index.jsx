import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PlaylistsPage from './Playlists';
import { fetchCategory } from '../../../../actions/browse';
import Category from '../../../../components/Category';
import Error from '../../../../components/Error';
import Loader from '../../../../components/Loader';

export class CategoryPage extends Component {
  static propTypes = {
    accessToken: PropTypes.string,
    category: PropTypes.shape({
      name: PropTypes.string,
    }),
    error: PropTypes.shape({
      message: PropTypes.string,
    }),
    fetchCategory: PropTypes.func,
    isLoading: PropTypes.bool,
    match: PropTypes.shape({
      params: PropTypes.shape({
        category_id: PropTypes.string,
      }),
      url: PropTypes.string,
    }),
  };

  static defaultProps = {
    accessToken: '',
    category: {
      name: '',
    },
    error: {
      message: '',
    },
    fetchCategory: () => {},
    isLoading: false,
    match: {
      params: {
        category_id: '',
      },
      url: '',
    },
  };

  componentDidMount() {
    const { accessToken, fetchCategory, match } = this.props;

    fetchCategory(accessToken, match.params.category_id);
  }

  render() {
    const {
      category, error, isLoading, match,
    } = this.props;

    return (
      <Switch>
        <Route path={`${match.url}/playlists`} component={PlaylistsPage} />
        <Route
          path={match.url}
          component={() => {
            if (error) {
              return <Error>{error.message}</Error>;
            }

            if (isLoading) {
              return <Loader />;
            }

            return (
              <div>
                <Helmet>
                  <title>{category.name}</title>
                </Helmet>
                <Category {...category} />
              </div>
            );
          }}
        />
      </Switch>
    );
  }
}

const mapStateToProps = state => ({
  ...state,
  category: state.browse.category,
  error: state.browse.error,
  isLoading: state.browse.isLoading,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    fetchCategory,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CategoryPage);
