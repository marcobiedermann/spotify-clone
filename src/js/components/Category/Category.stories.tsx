import type { Meta, StoryObj } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import Category from '.';
import categoryFixtures from './__fixtures__';

const meta: Meta<typeof Category> = {
  component: Category,
  args: categoryFixtures,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
};

type Story = StoryObj<typeof Category>;

const Default: Story = {};

export { Default };
export default meta;
