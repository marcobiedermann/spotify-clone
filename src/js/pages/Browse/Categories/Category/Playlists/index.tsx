import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { Error, Loader, Playlists } from '../../../../../components';
import { useBrowseCategoryPlaylists } from '../../../../../hooks/browse/categories';

function PlaylistsPage(): JSX.Element {
  const { categoryId } = useParams();
  const { data, error, isError, isLoading } = useBrowseCategoryPlaylists(categoryId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isLoading) {
    return <Loader />;
  }

  const { playlists } = data;

  return (
    <>
      <Helmet>
        <title>Playlists</title>
      </Helmet>
      <Playlists items={playlists.items} />
    </>
  );
}

export default PlaylistsPage;
