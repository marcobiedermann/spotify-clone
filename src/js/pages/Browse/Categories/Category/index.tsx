import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { Category, Error, Loader } from '../../../../components';
import { useBrowseCategory } from '../../../../hooks/browse/categories';

function CategoryPage(): JSX.Element {
  const { categoryId } = useParams();
  const { data, error, isError, isLoading } = useBrowseCategory(categoryId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isLoading) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Category {...data} />
    </>
  );
}

export default CategoryPage;
