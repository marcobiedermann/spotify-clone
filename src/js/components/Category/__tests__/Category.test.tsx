import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { expect, it } from 'vitest';
import Category from '..';
import categoryFixture from '../__fixtures__';

it('renders correctly', () => {
  const { container } = render(
    <MemoryRouter>
      <Category {...categoryFixture} />
    </MemoryRouter>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
