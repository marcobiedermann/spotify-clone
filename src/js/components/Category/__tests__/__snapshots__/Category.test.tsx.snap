// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<figure>
  <a
    href="/browse/categories//playlists"
  >
    <img
      alt=""
      height="10"
      src=""
      width="100"
    />
  </a>
  <figcaption>
    <a
      href="/browse/categories//playlists"
    />
  </figcaption>
</figure>
`;
